ENH: `HillClimbSearch` refactor (sklearn compatible)

XAheli opened the pull request:
### Issue number(s) that this pull request fixes

- Refs #2410

XAheli committed:
ENH: HillClimbSearch refactor (sklearn compatible) - closes #2410 partially

ankurankan committed:
Merge branch 'dev' into feature/refactor-hillclimb-sklearn

ankurankan committed:
Merge branch 'dev' into feature/refactor-hillclimb-sklearn

ankurankan commented on pgmpy/tests/test_causal_discovery/test_HillClimbSearch.py:
@XAheli Are these tests different from what we had earlier? I don't see any exact delta score comparison tests.

XAheli replied:
Yes I did not add some of the test methods from [test_estimators/test_HillClimbSearch.py](https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/tests/test_estimators/test_HillClimbSearch.py) `test_legal_operations`, `test_legal_operations_titanic`

Will rectify

XAheli committed:
fix: black  fixes

XAheli committed:
Merge branch 'feature/refactor-hillclimb-sklearn' of github.com:XAheli/pgmpy into feature/refactor-hillclimb-sklearn

ankurankan committed:
Extension template for causal discovery (#2521)

### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this
repository.

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix
branch** (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the **dev branch**
(left side). Also you should start *your branch* off *our dev*.
- [ ] Check the commit's or even all commits' message styles matches our
requested structure.

### Issue number(s) that this pull request fixes
- Fixes #

### List of changes to the codebase in this pull request
- 
-
-

ankurankan committed:
Merge branch 'feature/refactor-hillclimb-sklearn' of github.com:XAheli/pgmpy into pr/2495

ankurankan committed:
Minor fixes and refactor to use a score mixin class

ankurankan commented:
@XAheli Thanks for the PR.
